src/python/pants/goal/exec_stats.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import base64
+import logging
+import textwrap
+from collections import Counter
+from io import BytesIO
+from typing import cast
+
+from pants.engine.internals.scheduler import Workunit
+from pants.engine.rules import collect_rules, rule
+from pants.engine.streaming_workunit_handler import (
+    StreamingWorkunitContext,
+    WorkunitsCallback,
+    WorkunitsCallbackFactory,
+    WorkunitsCallbackFactoryRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.option.subsystem import Subsystem
+
+logger = logging.getLogger(__name__)
+
+
+class ExecStatsSubsystem(Subsystem):
+    options_scope = "exec-stats"
+    help = "An aggregator for Pants execution stats, such as cache metrics."
+
+    @classmethod
+    def register_options(cls, register):
+        register(
+            "--log",
+            advanced=True,
+            type=bool,
+            default=False,
+            help=(
+                "At the end of the Pants run, log all counter metrics and histograms, which give "
+                "metrics on Pants execution like the number of cache hits.\n\nFor histograms to "
+                "work, you must add `hdrhistogram` to `[GLOBAL].plugins`."
```

Comment:
Bike shed #3, should we split out the histograms from counters? `--counters` and `--histograms` as separate options.

Histograms are really chatty, and it might be nice to be able to disable those. (Although, you can already disable them by not adding `hdrhistogram` to your plugins.)

Reply:
I agree in splitting them out. The default histogram output can overwhelm the console especially as we add more observation metrics. This may also point at just dumping a few Pnn values (e.g., P25, P50, P75, P90, P95, P99) for the histograms by default and only doing the full dump via an option.

Reply:
Keeping them merged and rendering a few hardcoded quantile values would be my vote.

Reply:
I agree with showing a few hardcoded quantile values. This is too much info currently.

--

I'd still personally vote for separating them because it feels like it will scale better, e.g. if we add another type of metric or want to add `--histograms-verbosity` to complement `--histograms`. And for my personal use case, it's nice being able to ignore histograms most the time.

Finally, with a dedicated `--histograms`, we can error if the hdrhistograms library isn't installed, which is better than the warning before.

But I'm fine going either way.

Reply:
New types of metrics are not something that come along every decade... I wouldn't bias toward enabling that.